Add ecl command to list colors with descriptions

Ayushd785 opened the pull request:
Implements new `ecl` command that lists all color palette keys with human-readable explanations, similar to the `el` command for config variables.
   
   This addresses the issue where color names like `usrcmt`, `floc`, `fname` etc. are unclear without verbose explanation.
   
   
<img width="769" height="529" alt="image" src="https://github.com/user-attachments/assets/6e816c0b-4375-4ee4-a092-e4a98d831b7b" />

   
   **Changes:**
   - Added `description` field to color palette keys in `librz/cons/pal.c`
   - Added `rz_cons_pal_list_with_help()` function to display colors with descriptions
   - Added `ecl` command definition in YAML
   - Added command handler `rz_cmd_eval_color_list_help_handler()`
   - Added test case in `test/db/cmd/cmd_ec`
   
   Fixes #5462

Ayushd785 commented:
Hello @notxvilka ,

I hope this finds you well. I've implemented the `ecl` command as described in issue #5462. This is my first contribution to Rizin, so I would greatly appreciate your feedback and guidance on any improvements needed.

Thank you for your time.

wargio commented:
i wonder if this could be just part of the yaml file we use for generating the commands @notxvilka

Ayushd785 commented:
> i wonder if this could be just part of the yaml file we use for generating the commands @notxvilka

Hi @wargio! Thanks for the feedback. 

You're right to question this approach. I implemented it manually because I wasn't sure if the YAML generation system could handle:

1. Custom formatting (the padding/spacing in the output)
2. Accessing the color descriptions from the palette structure

Could you clarify if there's a way to achieve the same result using only YAML definitions? I'd be happy to refactor it if there's a cleaner approach!

notxvilka commented:
> i wonder if this could be just part of the yaml file we use for generating the commands @notxvilka

Not sure what do you mean?

wargio commented:
> > i wonder if this could be just part of the yaml file we use for generating the commands @notxvilka
> 
> Hi @wargio! Thanks for the feedback. 
> 
> You're right to question this approach. I implemented it manually because I wasn't sure if the YAML generation system could handle:
> 
> 1. Custom formatting (the padding/spacing in the output)
> 2. Accessing the color descriptions from the palette structure
> 
> Could you clarify if there's a way to achieve the same result using only YAML definitions? I'd be happy to refactor it if there's a cleaner approach!
> 

The yaml approach does not support colors but the printer will use colors for this.

@notxvilka the "details" feature of the yaml to c command generator could fit this usage, and just have name and comment.

See as example the command `dbC`

Ayushd785 commented:
> > > i wonder if this could be just part of the yaml file we use for generating the commands @notxvilka
> > 
> > 
> > Hi @wargio! Thanks for the feedback.
> > You're right to question this approach. I implemented it manually because I wasn't sure if the YAML generation system could handle:
> > 
> > 1. Custom formatting (the padding/spacing in the output)
> > 2. Accessing the color descriptions from the palette structure
> > 
> > Could you clarify if there's a way to achieve the same result using only YAML definitions? I'd be happy to refactor it if there's a cleaner approach!
> 
> The yaml approach does not support colors but the printer will use colors for this.
> 
> @notxvilka the "details" feature of the yaml to c command generator could fit this usage, and just have name and comment.
> 
> See as example the command `dbC`

shall i proceed with this approach then?

Ayushd785 commented:
> > > > i wonder if this could be just part of the yaml file we use for generating the commands @notxvilka
> > > 
> > > 
> > > Hi @wargio! Thanks for the feedback.
> > > You're right to question this approach. I implemented it manually because I wasn't sure if the YAML generation system could handle:
> > > 
> > > 1. Custom formatting (the padding/spacing in the output)
> > > 2. Accessing the color descriptions from the palette structure
> > > 
> > > Could you clarify if there's a way to achieve the same result using only YAML definitions? I'd be happy to refactor it if there's a cleaner approach!
> > 
> > 
> > The yaml approach does not support colors but the printer will use colors for this.
> > @notxvilka the "details" feature of the yaml to c command generator could fit this usage, and just have name and comment.
> > See as example the command `dbC`
> 
> Okay @wargio so, shall i proceed with this approach then?

Ayushd785 committed:
Add ecl command using YAML details feature

Implement new ecl command to list color palette keys with descriptions
using Rizin's YAML command generation system.

- Add ecl command definition in cmd_eval.yaml with details section
- Add handler function to iterate and print color names with descriptions
- Add test case for ecl command
- Use ^anchor in grep to avoid substring matching issues

Fixes #5462

Ayushd785 commented:
Hello ,@notxvilka  @wargio  ! I've updated the implementation based on the feedback. 

New approach: Now using YAML `details` feature in `cmd_eval.yaml` to define all color names and descriptions, with a handler that reads and displays them directly from the command descriptor. This integrates better with Rizin's command generation system without modifying `pal.c`.

The test now uses `^` anchor in grep to avoid substring matching issues.

Ayushd785 committed:
Fix padding in ecl command output

Adjust padding calculation from max_len+1 to max_len+2 to match
expected spacing in test output.

Ayushd785 commented:
Hii @notxvilka ,
Some CI checks were failing earlier ,I’ve tried fixing them now. Could you please re-run the CI checks?
I’m really sorry for the inconvenience. I’m still quite new to this organization, but I’ll make sure to follow best practices and improve with every contribution.

Ayushd785 commented:
@notxvilka, Sir, all tests are now passing successfully. I would be grateful for your review and any feedback or guidance you may have. Thank you for your time and patience with my first contribution.

notxvilka reviewed (approved):
Not particularly happy about the possibility of de-synchronization of the palette described in the code and command detail, but it's good enough for now.

Ayushd785 commented:
<img width="1919" height="1047" alt="image" src="https://github.com/user-attachments/assets/27fb851d-c5b7-40f2-9c39-2d4dfabc607b" />

It works perfectly

Ayushd785 commented:
> Not particularly happy about the possibility of de-synchronization of the palette described in the code and command detail, but it's good enough for now.

No problem, I can definitely work on it

wargio commented:
> Not particularly happy about the possibility of de-synchronization of the palette described in the code

this could be fixed with a custom linter checking the yaml vs the c code.

Ayushd785 commented:
> > Not particularly happy about the possibility of de-synchronization of the palette described in the code
> 
> this could be fixed with a custom linter checking the yaml vs the c code.

I will surely fix it sir

Ayushd785 committed:
Removed colon from ecl command output for cleaner format

Ayushd785 committed:
Add color visualization to ecl command output

Implement colored blocks (##) in ecl command output, similar to the ec command.
This provides visual representation of each color alongside its description.

Ayushd785 commented:
@notxvilka , @wargio I used Rizin's existing color palette system to get all 70 color names and their actual colors. For each color, I looked up its description from the YAML file we already created, converted the color to ANSI codes, and printed it in the format: [colored ##] colorname description.
Instead of duplicating all the color data, I just reused what's already in the codebase and combined it with our descriptions. This way if colors change in the future, ecl automatically stays in sync.

<img width="1235" height="1053" alt="image" src="https://github.com/user-attachments/assets/09e6d4d1-974d-4540-a060-65dffd92bdb1" />

The ecl command is in perfect sync with the ec command they both follow the same color scheme for each colorname

Ayushd785 commented:
I’ve used AI tools like Gemini 2.5 Pro (web version) to write more efficient code and brainstorm different approaches, and GitHub Copilot in VS Code to better understand the codebase.

wargio reviewed (approved):
If that screenshot is how it looks like with the latest changes, then i love it.

Ayushd785 commented:
> If that screenshot is how it looks like with the latest changes, then i love it.

I’m really glad to know that you liked it. Thank you so much for all the feedback and guidance!
I would love to contribute more.
